e3sm_diags/derivations/formulas_cosp.py

```diff
@@ -0,0 +1,435 @@
+
+    Returns
+    -------
+    xr.Dataset
+        The dataset with target variable, which is the standardized version
+        of the derived variable, ``var``.
+    """
+    ds_new = ds.copy()
+    prs = _get_cloud_axis(var, "prs")
+    tau = _get_cloud_axis(var, "tau")
+
+    # Mask on the min and max of prs and/or tau, then subset by dropping masked
+    # values.
+    var_std = _subset_cloud_var(var.copy(), prs, tau)
+
+    # Align the dataset dimensions with the standardized variabe. `xr.align`
+    # returns both objects and element 0 is the xr.Dataset that is needed.
+    ds_final = xr.align(ds_new, var_std)[0]
+    ds_final[target_var_key] = var_std
+    ds_final = ds_final.drop_vars(str(var.name))
+
+    # TODO: These functions don't actually add missing cloud bounds yet
+    # because it is based on the logic from the legacy code, which does not
+    # correctly add bounds if they are missing.
+    ds_final = _add_missing_cloud_bnds(ds_final, prs, "prs")
+    ds_final = _add_missing_cloud_bnds(ds_final, tau, "tau")
+
+    return ds_final
+
+
+def cosp_bin_sum(target_var_key: str, var: xr.DataArray) -> xr.DataArray:
+    """Get the cosp bin sum for the derived variable.
+
+    Parameters
+    ----------
+    target_var_key : str
+        The target variable key (e.g,. "CLDTOT_TAU1.3_ISCCP").
+    var : xr.DataArray
+        The variable in the dataset used for deriving the target variable
+        (e.g., "CLD_MODIS").
+
+    Returns
+    -------
+    xr.DataArray
+        The target variable, which is the cosp bin sum of the derived variable,
+        ``var``.
+    """
+    # 1. Get cloud axes
+    prs = _get_cloud_axis(var, "prs")
+    tau = _get_cloud_axis(var, "tau")
+
+    # 2. Get the prs and tau axis adjustment ranges if they are set.
+    prs_adj_range = CLOUD_BIN_SUM_MAP[target_var_key]["prs"]
+    tau_adj_range = CLOUD_BIN_SUM_MAP[target_var_key]["tau"]
+
+    # 3. Get prs range, lim.
+    prs_range = _get_prs_subset_range(prs, prs_adj_range)
+
+    # 4. Get tau range and lim.
+    tau_range, tau_lim = _get_tau_subset_range_and_str(tau, tau_adj_range)
+
+    # 4. Get the axes mask conditional and subset the variable on it.
+    cond = (
+        (prs >= prs_range[0])
+        & (prs <= prs_range[-1])
+        & (tau >= tau_range[0])
+        & (tau <= tau_range[-1])
+    )
+    var_sub = var.where(cond, drop=True)
+
+    # 5. Sum on axis=0 and axis=1 (tau and prs)
+    var_sum = var_sub.sum(dim=[prs.name, tau.name])
+
+    # 6. Set the variable's long name based on the original variable's name and
+    # prs ranges.
+    var_key = str(var.name)
+    simulation = _get_simulation_str(var_key)
+    prs_cloud_level = _get_prs_cloud_level_str(var_key, prs_range, prs_adj_range)
+
+    if simulation is not None and prs_cloud_level is not None:
+        var_sum.attrs["long_name"] = f"{simulation}: {prs_cloud_level} with {tau_lim}"
+
+    # 7. Convert units to %.
+    final_var = convert_units(var_sum, "%")
```

Comment:
This replaces the `convert_units` calls in `derivations.py`.

Reply:
Nice clean-up!